[Sema] Extend callee diagnosis to complex args including generics, e.g. (Void) -> T

gregomni opened the pull request:
Correctly determine callee closeness for func/ops that include generics as part of more complicated parameters, i.e. tuple or closure args containing generics as elements or args/results. Still only handling single archetypes.

Tests added, all tests pass. The one changed diagnostic in misc_diagnostics is sort of a regression, although both messages are of about equal quality, I think. This commit lets evaluateCloseness find a second partial match for ~= there, but the two partial matches involve different parameter archetypes, so filtering makes the callee argument failure info invalid as a result, thus going back to the generic message as opposed to a specific argument failure message. (In short, the older specific error was actually incorrectly over-specific.)

slavapestov commented on lib/Sema/CSDiag.cpp:
```diff
@@ -980,6 +981,69 @@ static bool argumentMismatchIsNearMiss(Type argType, Type paramType) {
+  if (paramType->getKind() != actualArgType->getKind())
+    return false;
+  
+  // A little walker that just collects direct child types. Can't easily have a
```
Take a look at include/swift/AST/TypeMatcher.h, it might be what you need here

gregomni replied:
Aha! Thank you, that's exactly what I needed. It seemed like something like that must exist, but somehow I missed finding it.

slavapestov commented on lib/Sema/CSDiag.cpp:
```diff
@@ -1064,16 +1128,24 @@ CalleeCandidateInfo::evaluateCloseness(Type candArgListType,
-      if (auto genericParam = paramType->getAs<GenericTypeParamType>())
-        paramType = genericParam->getDecl()->getArchetype();
-      if (paramType->is<ArchetypeType>() && !rArgType->hasTypeVariable()) {
+      if (matched && archetypes.size() == 1 && !rArgType->hasTypeVariable()) {
```
hasTypeVariable() means the type contains a GenericTypeParamType or a DependentMemberType. I don't believe that can happen here since these are contextual types (written in terms of archetypes) and not interface types (written in terms of type parameters).

Maybe try changing that to an assert. If you're seeing both interface and contextual types here, there might be something else to look at.

gregomni replied:
Actually, I think you mean hasTypeParameter(). hasTypeVariable() is for if the type contains a TypeVariableType that gets inserted by the constraint system. By the time the solver is done there should be none left, but since this is in the couldn't-solve-it path, they can still exist.

This seems rare, but it does happen, like in test/Constraints/diagnostics.swift:644:
```
func foo() -> [Int] {
  return Array <Int> (count: 1) { // expected-error {{cannot invoke initializer for type 'Array<Int>' with an argument list of type '(count: Int, () -> _)'}}
    // expected-note @-1 {{expected an argument list of type '(count: Int, repeatedValue: Element)'}}
    return 1
  }
}
```
where rArgType (when checking the trailing closure param) is:
```
(function_type
  (input=tuple_type num_elements=0)
  (output=type_variable_type id=1))
```

slavapestov replied:
Yes, I meant hasTypeParameter(), thanks.

slavapestov commented on lib/Sema/CSDiag.cpp:
```diff
@@ -980,6 +981,69 @@ static bool argumentMismatchIsNearMiss(Type argType, Type paramType) {
+      return false;
+  }
+  
+  // Unfortunate special case, FunctionType also needs to compare ExtInfo. This is the only
```
Do you have a test for this case?

gregomni replied:
Without this check, test/attr/attr_autoclosure.swift:53 regresses and emits the old worse diagnoses from before https://github.com/apple/swift/pull/1042:
```
func overloadedEach<O: P1>(source: O, _ closure: () -> ()) {
}

func overloadedEach<P: P2>(source: P, _ closure: () -> ()) {
}

struct S : P2 {
  typealias Element = Int
  func each(@autoclosure closure: () -> ()) {
    overloadedEach(self, closure) // expected-error {{invalid conversion from non-escaping function of type '@autoclosure () -> ()' to potentially escaping function type '() -> ()'}}
  }
}
```

slavapestov commented on lib/Sema/CSDiag.cpp:
```diff
@@ -980,6 +981,69 @@ static bool argumentMismatchIsNearMiss(Type argType, Type paramType) {
+  
+  if (paramType->isEqual(actualArgType))
+    return true;
+  if (paramType->getKind() != actualArgType->getKind())
```
I think that's too conservative, what if paramType is a type alias or some other sugared form of actualArgType? Perhaps try to come up with a test case for that, and see if getCanonicalType() helps

gregomni replied:
Good direction for next commit. :)  I'm being conservative because failing to match just leads to the existing generic couldn't-call-with-these-args diagnosis, and I don't ever want to accidentally get too liberal, because then we'll emit a diagnosis which is more specific but possibly wrong.

slavapestov commented on lib/Sema/CSDiag.cpp:
```diff
@@ -980,6 +981,69 @@ static bool argumentMismatchIsNearMiss(Type argType, Type paramType) {
+                                     SmallVector<ArchetypeType *, 4> &archetypes,
+                                     SmallVector<Type, 4> &substitutions) {
+  if (auto genericParam = paramType->getAs<GenericTypeParamType>())
+    paramType = genericParam->getDecl()->getArchetype();
```
What if the paramType is a DependentMemberType? (Ie, an associated type reference).

Also getArchetype() is going away -- try this instead, which addresses both issues:

if (paramType->isTypeParameter())
    paramType = ArchetypeBuilder::mapTypeIntoContext(dc, paramType, nullptr);

gregomni replied:
I'll look into this, thanks!

slavapestov commented on lib/Sema/CSDiag.cpp:
```diff
@@ -980,6 +981,69 @@ static bool argumentMismatchIsNearMiss(Type argType, Type paramType) {
+    if (!argFunc || paramFunc->getExtInfo() != argFunc->getExtInfo())
+      return false;
+  }
+  return true;
```
Suppose findGenericSubstitutions() is called with paramType ```B<T>``` where T has no constraints, and argType ```C<Int>```. Here we have two types that have the same kind, same number of children, and the substitution Int -> T is valid. However they don't match.

Just as with function types, BoundGenericType has 'extra data' beyond its child types that needs to be compared.

Maybe if the TypeWalker utility is not appropriate here, you should just destructure types by hand -- there are not that many cases, I think this covers it:

- bound generic types -- compare the decl and arguments
- function types -- compare input, output and extinfo
- metatype types -- compare instance types
- tuple types -- compare elements
- archetypes -- record them in the substitution list
- anything else -- check for equality

gregomni replied:
You're right, I need to check for that. Looks like this hits that bad case:
```
struct A<T> {
  let t: T
  init(_ t: T) {
    self.t = t
  }
}

struct B<T> {
  
}

func foo<T>(a: A<T>) -> T {
  return a.t
}

func foo<T>(b: B<T>) -> Double {
  return 0
}

let a = A(3)
let d: Double = foo(a)
```
(It needs multiple generics to choose among to reach this point, otherwise it figures out at the expression level that an A can't be a B.)

I'll work on this!

I'll take a look at converting to using TypeMatcher and see how the complexity falls out. Doing the destructure by hand is probably more safely conservative, (and how I initially attacked this), but the result just looks really ugly.

slavapestov commented on test/Constraints/diagnostics.swift:
```diff
@@ -82,6 +83,15 @@ let a = A()
 for j in i.wibble(a, a) { // expected-error {{type 'A' does not conform to protocol 'SequenceType'}}
 }
 
+// Generic as part of function/tuple types
```
Add tests involving typealiases and ensure the diagnostics are still useful and we don't fall back to the general path, and that the diagnostic involves the sugared typealias type and not what it points to

slavapestov commented:
I'm not very familiar with the diagnostics code so I'm still learning too. I noticed a few potential improvements. I won't insist you address all of them in this patch, but take a look and see if any look like low-hanging fruit. The code can always be improved later too.

gregomni commented:
I'd like to try the TypeMatcher path and also fix the differing bound generic structs case, and then leave matching with type aliases and other sugared forms for the (hopefully near) future.

gregomni commented:
Ok! Rewrote this patch to use TypeMatcher, which is a much cleaner way to do it.

- Switched Generic -> Archetype transform to use ArchetypeBuilder::mapTypeIntoContext() as you suggested. So this now is hopefully future-proofed, and it also should work for dependent member types.

- TypeMatcher had existing mismatch code for checking the base of BoundGenericType types, so that potential problem got fixed by adopting it.

- Noticed that one of the improved diagnoses from one of my earlier patches here regressed with the TypeMatcher, and when I looked into it, it was because the previous code got a correct result accidentally by doing something wrong. The parameter was a SubstitutedType and the previous code was looking inside its children (its original type) and finding an archetype there that didn't match. That no longer happens with TypeMatcher because the paramType arg is canonicalized, which replaces the SubstitutedType with its replacement type.

The correct solution to do this in general and not by accident is to pre-populate the substitution vectors with any substitutions that have already been made, so this code also does that in findGenericSubstitutions() now, which leads to a few more corrected diagnoses (such as in Generics/function_defs.swift).
  
- Still not looking for type aliases or sugared types, that's for the future.

All tests pass, new test added, changes to diagnoses for existing tests also cover the new features here. Went ahead and squashed right away because the original commit got entirely changed anyway, so there wasn't anything readable about the diff between the two commits.

slavapestov commented:
This looks good! I'll merge it once it goes through CI.

slavapestov commented:
@swift-ci Please test

gregomni commented:
Oops. I see I was running just -t instead of -T and a couple of the validation tests failed. Will go back and look at those.

gregomni commented:
@swift-ci Please test

gregomni commented:
@slavapestov Looks like only people with commit access can request testing. Made a second commit to check for UnresolvedType in the paramType and changed to a better diagnosis in one of the validation tests, and everything now passes for me locally.

slavapestov commented:
@swift-ci Please test

slavapestov commented:
@gregomni Requesting tests doesn't always work for me either -- I think it might just be flaky. I'll follow up internally

gregomni commented:
@slavapestov I went and found the announcement and it does explicitly say that only committers can kick it off: <https://swift.org/blog/swift-ci/>. I can understand that, since it's a fairly long-lived process you don't want random people kicking off zillions of tests.

gregomni commented:
Huh. There is one failure in a test in IRGen, which I don't see how it could be related to these changes.

modocache commented:
@gregomni I'm curious: could you post the output of the failing test in IRGen?

gregomni commented:
For what it's worth, the next build on that jenkins instance <https://ci.swift.org/job/swift-PR-osx/18/>, failed with the exact same IRGen error, and it was just doing a build for stdlib changes. So I think that test must be broken by some previous pull.

slavapestov commented:
I belive the broken test was a commit by @DougGregor that he has since fixed.

slavapestov commented:
@swift-ci Please test

slavapestov commented on lib/Sema/CSDiag.cpp:
```diff
@@ -1325,7 +1397,8 @@ void CalleeCandidateInfo::filterList(ArrayRef<CallArgParam> actualArgs) {
     if (!inputType) return { CC_GeneralMismatch, {}};
-    return evaluateCloseness(inputType, actualArgs);
+    Decl *decl = candidate.getDecl();
+    return evaluateCloseness(decl ? decl->getInnermostDeclContext() : nullptr, inputType, actualArgs);
```
Last nitpick before I merge this -- this line is too long, wrap it after the nullptr

gregomni commented:
Sure thing. Wrapped that line, and squashed.

jtbandes commented:
Cool!
